### PR TITLE
[ppc] poll interruption checkpoint in generic trampoline

### PR DESCRIPTION
**Pull request by lewurm**
fixes https://github.com/mono/mono/issues/7396

@NattyNarwhal could you please test this patch on a big endian machine? There is a case where I'm unsure about function descriptor stuff.

### CODE REVIEW

**Comment by NattyNarwhal** [comment]
exceptions test passes. Unfortunately, iltest has [regressed](https://gist.github.com/NattyNarwhal/c83707ee14bcf72247859da111ebeab1), but I'm not sure what caused that.

**Comment by NattyNarwhal** [comment]
Testing on Linux seems to put #7430 at fault; this LGTM

**Comment by NattyNarwhal** [comment]
so the tl;dr this might not seem to not handle calling conventions on PPC64BE Linux (different kind of error) but AIX and PPC64LE Linux are fine. Something about the link register?

Pleased it works here though!

**Comment by lewurm** [comment]
@monojenkins build Linux PPC64 little endian

**Comment by NattyNarwhal** [comment]
FWIW, a gdb session on `ppc64be/linux`:

```
[0xfffb7e012b0: 0.20074 3] LEAVE: (wrapper runtime-invoke) <Module>:runtime_invoke_void__this___object_object (object,intptr,intptr,intptr)[OBJECT:(nil)]
[New Thread 0xfffb77af1c0 (LWP 16462)]
[New Thread 0xfffb6fff1c0 (LWP 16461)]

Program received signal SIGSEGV, Segmentation fault.
0x00000fffb79101d4 in ?? ()
(gdb) i r
r0             0xfffb79101d4    17590970810836
r1             0xfffffffe460    17592186037344
r2             0x1047b628       273135144
r3             0x0      0
r4             0x5a9502 5936386
r5             0xfffb7000638    17590961309240
r6             0xfffffffe0b0    17592186036400
r7             0xffffffffffffffff       18446744073709551615
r8             0x1049ee90       273280656
r9             0x0      0
r10            0x1049f418       273282072
r11            0x1      1
r12            0xfffb7e23380    17590976131968
r13            0xfffb7e089c0    17590976022976
r14            0x1      1
r15            0xffffffff7c5    17592186042309
r16            0x161129ff       370223615
r17            0x2d     45
r18            0x0      0
r19            0x0      0
r20            0x1038fe08       272170504
r21            0x104758e8       273111272
r22            0x10513690       273757840
---Type <return> to continue, or q <return> to quit---
r23            0xfffffffeae0    17592186039008
r24            0x104a6260       273310304
r25            0x104a38a0       273299616
r26            0x10513690       273757840
r27            0x0      0
r28            0x0      0
r29            0xfffffffeb80    17592186039168
r30            0x1043b928       272873768
r31            0xfffffffe7f0    17592186038256
pc             0xfffb79101d4    0xfffb79101d4
msr            0x800000000000d032       9223372036854829106
cr             0x22002422       570434594
lr             0xfffb79101d4    0xfffb79101d4
ctr            0xfffb7e23380    17590976131968
xer            0x20000002       536870914
orig_r3        0xc000000000003554       -4611686018427374252
trap           0x300    768
(gdb) bt
#0  0x00000fffb79101d4 in ?? ()
#1  0x00000fffb5344478 in ?? ()
#2  0x00000fffb5344660 in ?? ()
#3  0x000000001003760c in mono_jit_runtime_invoke (method=<optimized out>,
    obj=0x0, params=0xfffffffeae0, exc=<optimized out>, error=0xfffffffeb80)
    at mini-runtime.c:2875
#4  0x0000000010222a18 in do_runtime_invoke (method=0x104a7788,
    obj=<optimized out>, params=<optimized out>, exc=<optimized out>,
    error=0xfffffffeb80) at object.c:2842
#5  0x0000000010225c04 in do_exec_main_checked (method=0x104a7788,
    args=<optimized out>, error=0xfffffffeb80) at object.c:4755
#6  0x0000000010083a40 in mono_jit_exec (domain=<optimized out>,
    assembly=<optimized out>, argc=<optimized out>, argv=0xffffffff408)
    at driver.c:1202
#7  0x0000000010085f68 in main_thread_handler (user_data=<synthetic pointer>)
    at driver.c:1279
#8  mono_main (argc=<optimized out>, argv=<optimized out>) at driver.c:2380
#9  0x000000001002f2a4 in mono_main_with_options (argv=0xffffffff3f8, argc=3)
    at main.c:50
#10 main (argc=3, argv=0xffffffff3f8) at main.c:398
(gdb) p mono_pmip (0x00000fffb79101d4)
$1 = 0x10514c60 "<0xfffb79101d4 - generic_trampoline_jit trampoline>"
(gdb) x/32i $pc - 32
   0xfffb79101b4:       ori     r12,r12,0
   0xfffb79101b8:       rldicr  r12,r12,32,31
   0xfffb79101bc:       oris    r12,r12,4164
   0xfffb79101c0:       ori     r12,r12,62536
   0xfffb79101c4:       ld      r2,8(r12)
   0xfffb79101c8:       ld      r12,0(r12)
   0xfffb79101cc:       mtlr    r12
   0xfffb79101d0:       blrl
=> 0xfffb79101d4:       ld      r2,8(r3)
   0xfffb79101d8:       ld      r3,0(r3)
   0xfffb79101dc:       mr      r14,r3
   0xfffb79101e0:       lis     r12,0
   0xfffb79101e4:       ori     r12,r12,0
   0xfffb79101e8:       rldicr  r12,r12,32,31
   0xfffb79101ec:       oris    r12,r12,4166
   0xfffb79101f0:       ori     r12,r12,15152
   0xfffb79101f4:       ld      r2,8(r12)
   0xfffb79101f8:       ld      r12,0(r12)
   0xfffb79101fc:       mtlr    r12
   0xfffb7910200:       blrl
   0xfffb7910204:       cmpdi   r3,0
   0xfffb7910208:       beq     0xfffb7910248
   0xfffb791020c:       ld      r14,192(r1)
---Type <return> to continue, or q <return> to quit---
   0xfffb7910210:       ld      r1,0(r1)
   0xfffb7910214:       ld      r12,16(r1)
   0xfffb7910218:       mtlr    r12
   0xfffb791021c:       lis     r12,0
   0xfffb7910220:       ori     r12,r12,0
   0xfffb7910224:       rldicr  r12,r12,32,31
   0xfffb7910228:       oris    r12,r12,4167
   0xfffb791022c:       ori     r12,r12,24472
   0xfffb7910230:       ld      r2,8(r12)
```

Applying some Kremlinology, I believe `mono/mini/tramp-ppc.c` L392 is where it hangs at.

**Comment by lewurm** [comment]
@NattyNarwhal funny, I have the exact same addresses (on your machine, but a different build). disabling ASLR for the win 😉 

the crash happens in that part of code: https://github.com/lewurm/mono/blob/d49fb510e05aa4767651893aa58b12a58bbabf75/mono/mini/tramp-ppc.c#L372-L377

Specifically, this instance of trampoline is calling `mono_magic_trampoline ()`, thus the value in `r3` is the return value of that function which should be a function pointer/descriptor. The interesting thing now: If you put an assert that `mono_magic_trampoline` like `g_assert (ret);`, it _never_ triggers.

So that's really weird. I tried to single step after the last known good call from `mono_magic_trampoline`, but it was hard to follow. Though, we end up in this trampoline again _without_ calling `mono_magic_trampoline` which, if you believe the control flow of the assembly, should be impossible. I believe that the return address is left somewhere on the frame (wrongly), and something else picks it up later.

**Comment by NattyNarwhal** [comment]
the symbol in r12 is?

```
(gdb) info symbol 0xfffb7e23380
.__pthread_getspecific in section .text of /lib64/power6/libpthread.so.0
```

that doesn't look right....

**Comment by NattyNarwhal** [comment]
more g_print:

```
** Tramp ** addr 0x1044f458 type 0
** Tramp ** addr 0x1044f458 type 1
** Tramp ** addr 0x1044f398 type 2
** Tramp ** addr 0x1044f470 type 3
** Tramp ** addr 0x1044f3c8 type 4
** Tramp ** addr 0x1044f4a0 type 5
** Tramp ** addr 0x1044f380 type 6
** Tramp ** addr 0x1044f488 type 7
```

```
(gdb) info symbol mono_get_addr_from_ftnptr(0x1044f458)
.mono_magic_trampoline in section .text of /home/calvin/mono/mono/mini/mono-sgen
(gdb) info symbol 0x1044f458
mono_magic_trampoline in section .opd of /home/calvin/mono/mono/mini/mono-sgen
```

hmhmhmhmhmh

**Comment by lewurm** [comment]
`r12` isn't callee-saved, so seeing `__pthread_getspecific` there is "fine" (it could give us some hints where it's coming from though).

if you look at this part:
```
   0xfffb79101b4:       ori     r12,r12,0      // r12 = 0
   0xfffb79101b8:       rldicr  r12,r12,32,31  // r12 = (r12 << 32) & ((1 << 32) - 1) = 0
   0xfffb79101bc:       oris    r12,r12,4164   // r12 = 0 | (4164 << 16) = 0x10440000
   0xfffb79101c0:       ori     r12,r12,62536  // r12 = 0x1044000 | 62536 = 0x1044f448
   0xfffb79101c4:       ld      r2,8(r12)
   0xfffb79101c8:       ld      r12,0(r12)
   0xfffb79101cc:       mtlr    r12
   0xfffb79101d0:       blrl
```

so this points to the function descriptor of `mono_magic_trampoline`.

**Comment by lewurm** [comment]
soo, ppc64le does **not** work, as correctly caught by CI:

```
% gdb -q --args ./mono-sgen --regression objects.exe
Reading symbols from ./mono-sgen...done.
(gdb) r
Starting program: /home/lewurm/work/mono/mono/mini/mono-sgen --regression objects.exe
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/powerpc64le-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff6fff160 (LWP 51181)]
[New Thread 0x7ffff51ef160 (LWP 51182)]
Test run: image=/home/lewurm/work/mono/mono/mini/objects.exe, opts=

Thread 1 "mono-sgen" received signal SIGSEGV, Segmentation fault.
mono_find_jit_info_ext (domain=0x1048e550, jit_tls=0x104ea8a0, prev_ji=0x0, ctx=0x7fffffffd488, new_ctx=0x7fffffffd8a8, trace=0x0, lmf=0x7fffffffd210, save_locations=0x0,
    frame=0x7fffffffd290) at mini-exceptions.c:607
607                     *lmf = (MonoLMF *)(((gsize)(*lmf)->previous_lmf) & ~(SIZEOF_VOID_P -1));
(gdb) bt
#0  mono_find_jit_info_ext (domain=0x1048e550, jit_tls=0x104ea8a0, prev_ji=0x0, ctx=0x7fffffffd488, new_ctx=0x7fffffffd8a8, trace=0x0, lmf=0x7fffffffd210, save_locations=0x0,
    frame=0x7fffffffd290) at mini-exceptions.c:607
#1  0x00000000100bd7c8 in unwinder_unwind_frame (unwinder=0x7fffffffd300, domain=0x1048e550, jit_tls=0x104ea8a0, ctx=0x7fffffffd488, new_ctx=0x7fffffffd8a8,
    lmf=0x7fffffffd210, save_locations=0x0, frame=0x7fffffffd290, trace=0x0, prev_ji=0x0) at mini-exceptions.c:717
#2  0x00000000100c04e4 in handle_exception_first_pass (catch_frame=0x7fffffffd220, non_exception=<optimized out>, out_prev_ji=<synthetic pointer>, out_ji=<synthetic pointer>,
    out_filter_idx=<synthetic pointer>, obj=0x7ffff70038a8, ctx=0x7fffffffd488) at mini-exceptions.c:1680
#3  mono_handle_exception_internal (ctx=0x7fffffffdc18, obj=0x7ffff70038a8, resume=<optimized out>, out_ji=0x0) at mini-exceptions.c:2001
#4  0x000000001011d6b0 in mono_ppc_throw_exception (exc=0x7ffff70038a8, eip=<optimized out>, esp=<optimized out>, int_regs=<optimized out>, fp_regs=<optimized out>,
    rethrow=<optimized out>) at exceptions-ppc.c:349
#5  0x00007ffff7811fb8 in ?? ()
#6  0x00007ffff762530c in ?? ()
#7  0x00007ffff7624e68 in ?? ()
#8  0x0000000010081b04 in mini_regression_step (image=0x10491820, verbose=0, total_run=0x7fffffffe60c, total=0x7fffffffe620, opt_flags=<optimized out>, timer=0x104fd5b0,
    domain=0x1048e550) at driver.c:404
#9  0x0000000010083fd0 in mini_regression (total_run=0x7fffffffe60c, verbose=<optimized out>, image=0x10491820) at driver.c:525
#10 mini_regression_list (images=<optimized out>, count=<optimized out>, verbose=0) at driver.c:553
#11 mono_main (argc=<optimized out>, argv=<optimized out>) at driver.c:2292
#12 0x000000001002c900 in mono_main_with_options (argv=0x7fffffffed08, argc=3) at main.c:50
#13 main (argc=3, argv=0x7fffffffed08) at main.c:398
```

why didn't I notice that earlier? I developed this patch by compiling the runtime with `-O0`, CI runs with `-O2`. This hints very much to an ABI violation in this PR. Will investigate.

**Comment by NattyNarwhal** [comment]
Even with little-endian not working; the fact it (seemingly) works on AIX anyways is also curious, seeing as the ppc64be calling conventions on Linux are basically identical to AIX. (The #ifdef is even called `_CALL_AIX`, the opposite of `_CALL_ELF`!)

**Comment by NattyNarwhal** [comment]
I'm wonder, did you ever manage to make any headway on this? I'm using this PR on AIX and it's fine, but that of course doesn't matter when Linux has issues still.

**Comment by lewurm** [comment]
sorry, haven't gotten around to debug it yet. but I haven't forgotten about it 🙂

**Comment by lewurm** [comment]
@monojenkins Linux PPC64 little endian

**Comment by lewurm** [comment]
LMF was messed up if the generic trampoline was taking the interrupting path. not sure why it was only a problem on `-O2`, but it should be fixed now.

**Comment by lewurm** [comment]
@monojenkins build Linux PPC64 little endian

**Comment by lewurm** [comment]
@monojenkins build Linux PPC64 little endian
